benchmarks/fixed_pool_stress_test.cpp

```diff
@@ -0,0 +1,149 @@
+#include <iostream>
+#include <chrono>
+#include <string>
+#include <random>
+
+#include "umpire/config.hpp"
+
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Allocator.hpp"
+#include "umpire/strategy/FixedPool.hpp"
+
+#define CONVERT 1000000 //convert sec (s) to microsec (us)
+
+const int NUM_RND = 1000; //number of rounds (used to average timing)
+const int NUM_ALLOC = 512; //number of allocations used for testing
+const int FACTOR = 2048; //Default size used when creating FixedPool (64*sizeof(int)*8)
+
+////////////////////////////////////////////////////////////////////
+//Note: the total size of allocated memory is the multiplication of
+//SIZE and FACTOR, the default objects_per_pool value of 2048.
+////////////////////////////////////////////////////////////////////
+
+/*
+ * This function measures the time it takes to do NUM_ALLOC allocations and 
+ * then do NUM_ALLOC deallocations in the same order. The time is averaged across NUM_RND rounds. 
+ */
+void same_order(umpire::Allocator alloc, int SIZE)
+{
+  double time[2] = {0.0, 0.0};
+  void* allocations[NUM_ALLOC];
+
+  for(int i = 0; i < NUM_RND; i++) {
+    auto begin_alloc = std::chrono::system_clock::now();
+    for (int j = 0; j < NUM_ALLOC; j++)
+      allocations[j] = alloc.allocate(SIZE);
+    auto end_alloc = std::chrono::system_clock::now();
+    time[0] += std::chrono::duration<double>(end_alloc - begin_alloc).count()/NUM_ALLOC;
+
+    auto begin_dealloc = std::chrono::system_clock::now();
+    for (int h = 0; h < NUM_ALLOC; h++)
+      alloc.deallocate(allocations[h]);
+    auto end_dealloc = std::chrono::system_clock::now();
+    time[1] += std::chrono::duration<double>(end_dealloc - begin_dealloc).count()/NUM_ALLOC;
+  }
+
+  alloc.release();
```

Comment:
Okay, I promise to pay (better) attention this time.  Why is this call needed again?  You've appear to have done a matching allocate/deallocate pair above.

Reply:
When using the Google Benchmark stuff with FixedPool, we came across the problem where we would "deallocate" memory from a pool, but it was still marked as being 'used'. So even with deallocating, it wasn't "releasing" that memory to be used again by the next benchmark iterations. So the memory piled up and we kept running out of memory before the benchmark could finish even though the system memory info on Lassen says it should work fine. I just kind of rolled that logic over into the stress tests too (I don't think I've run the test without the release to see if it would be ok, I just assumed it was good to use for the stress tests too).

Reply:
@davidbeckingsale, please feel free to correct me if I'm wrong, but I think this should work without having to call `release()`.

In general both our DynamicPools and FixedPool implementations work by performing large allocations, called `blocks`, from the resource that they are a pool for (DEVICE, HOST, etc.).  The Pool will then perform a 'sub-allocation' from these `blocks`.  When the pool cannot find enough available memory in its set of `blocks`, it will perform another large allocation from the resource and create a new block and then return a sub-allocation from it to the caller.

When the application gives memory back to the pool by calling deallocate, the pool returns that memory to a portion of its associated `block`.  Once a `block` gets all of its memory returned, it "may" be coallesced into a completely free (available) block.  Until this time, the `block` will only be partially available to the application.

There are two `block` administrative things that may be done by the application:

- `release()` - This will return all of the completely available blocks to the resource.
- `coalesce()` - This will release all available `blocks` and then allocate a single larger block that is equivalent in size 
to the sum of all of those released `block` sizes.

It would be interesting to me to see the reproducer for the problem @kab163 is describing.

Reply:
Since this benchmark creates multiple pool allocators, it must `release` all of the memory resources used by the allocator when it is finished testing it before creating the next allocator to be tested.